stock_reserve_sale/model/sale.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2013 Camptocamp SA - Guewen Baconnier
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models
+from odoo.exceptions import UserError
+from odoo.osv import expression
+from odoo.tools.translate import _
+
+_LINE_KEYS = ["product_id", "product_uom_qty"]
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.depends(
+        "state", "order_line.reservation_ids", "order_line.is_stock_reservable"
+    )
+    def _compute_stock_reservation(self):
+        for sale in self:
+            has_stock_reservation = False
+            is_stock_reservable = False
+            for line in sale.order_line:
+                if line.reservation_ids:
+                    has_stock_reservation = True
+                if line.is_stock_reservable:
+                    is_stock_reservable = True
+            if sale.state not in ("draft", "sent"):
+                is_stock_reservable = False
+            sale.is_stock_reservable = is_stock_reservable
+            sale.has_stock_reservation = has_stock_reservation
+
+    has_stock_reservation = fields.Boolean(
+        compute="_compute_stock_reservation",
+        readonly=True,
+        multi="stock_reservation",
+        store=True,
+        string="Has Stock Reservations",
+    )
+    is_stock_reservable = fields.Boolean(
+        compute="_compute_stock_reservation",
+        readonly=True,
+        multi="stock_reservation",
+        store=True,
+        string="Can Have Stock Reservations",
+    )
+    reserves_count = fields.Integer(compute="_compute_reserves_count")
+    all_lines_reserved = fields.Boolean(
+        compute="_compute_reserves_count", store=True, default=False
+    )
+
+    def release_all_stock_reservation(self):
+        line_ids = [line.id for order in self for line in order.order_line]
+        lines = self.order_line.browse(line_ids)
+        lines.release_stock_reservation()
+        return True
+
+    def _compute_reserves_count(self):
+        reserve_ids = self.env["stock.reservation"]._read_group(
+            domain=expression.AND(
+                [
+                    [("sale_id", "in", self.ids)],
+                ]
+            ),
+            fields=["sale_id"],
+            groupby=["sale_id"],
+        )
+        lines = self.order_line.filtered(
+            lambda l: l.product_id.detailed_type != "service"
+        )
+        for order in self:
+            if reserve_ids:
+                order.reserves_count = reserve_ids[0]["sale_id_count"]
+            else:
+                order.reserves_count = 0
+            if order.reserves_count == len(lines):
```

Comment:
How is this possible?
That can not happens in a multi-recordset computation.
If this compute is called on a recordset with more than one `sale.order`, this will be never True.